src/input_stream.rs

```diff
@@ -341,6 +351,221 @@ impl<T: IsA<InputStream> + IsA<Seekable>> io::Seek for InputStreamRead<T> {
+                    .read_async_future(buffer, Priority::default());
+                self.state = State::Reading { pending };
+            }
+            State::Reading { .. } => {}
+            _ => panic!("Invalid state"),
+        };
+
+        self.state.get_pending()
+    }
+
+    fn get_data(&self) -> Poll<io::Result<&[u8]>> {
+        if let State::HasData {
+            ref buffer,
+            valid: (i, j),
+        } = self.state
+        {
+            return Poll::Ready(Ok(&buffer[i..j + 1]));
+        }
+        panic!("Invalid state")
+    }
+
+    fn set_waiting(&mut self, buffer: Vec<u8>) {
+        match self.state {
+            State::Reading { .. } | State::Transitioning => self.state = State::Waiting { buffer },
+            _ => panic!("Invalid state"),
+        }
+    }
+
+    fn set_has_data(&mut self, buffer: Vec<u8>, valid: (usize, usize)) {
+        match self.state {
+            State::Reading { .. } | State::Transitioning { .. } => {
+                self.state = State::HasData { buffer, valid }
+            }
+            _ => panic!("Invalid state"),
+        }
+    }
+
+    fn poll_fill_buf(&mut self, cx: &mut Context) -> Poll<Result<&[u8], futures::io::Error>> {
+        match self.state {
+            State::Failed(kind) => Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::from(kind),
+                BufReadError::Failed,
+            ))),
+            State::HasData { .. } => self.get_data(),
+            State::Transitioning => panic!("Invalid state"),
+            State::Waiting { .. } | State::Reading { .. } => {
+                let pending = self.set_reading();
+                match pending.poll_unpin(cx) {
+                    Poll::Ready(Ok((buffer, res))) => {
+                        if res == 0 {
+                            self.set_waiting(buffer);
+                            Poll::Ready(Ok(&[]))
+                        } else {
+                            self.set_has_data(buffer, (0, res - 1));
+                            self.get_data()
+                        }
+                    }
+                    Poll::Ready(Err((_, err))) => {
+                        let kind = err.kind::<crate::IOErrorEnum>().unwrap();
+                        self.state = State::Failed(kind);
+                        Poll::Ready(Err(io::Error::new(io::ErrorKind::from(kind), err)))
+                    }
+                    Poll::Pending => Poll::Pending,
+                }
+            }
+        }
+    }
+
+    fn consume(&mut self, amt: usize) {
+        if amt == 0 {
+            return;
+        }
+
+        if let State::HasData { .. } = self.state {
+            let has_data = mem::replace(&mut self.state, State::Transitioning);
+            if let State::HasData {
+                buffer,
+                valid: (i, j),
+            } = has_data
+            {
+                let available = j - i + 1;
+                let remaining = available - amt;
```

Comment:
Should this panic if more data is tried to be consumed than is currently available? What do the docs say?

Reply:
```
this function may do odd things if poll_fill_buf isn't called before calling it.
The amt must be <= the number of bytes in the buffer returned by poll_fill_buf.
```
So you're on your own if you don't play by the rule. Do you want an explicit panic?

Reply:
Sounds like a good idea

Reply:
Done. All your comments should be fixed now.